docs/cody/clients/model-configuration.mdx

```diff
@@ -0,0 +1,520 @@
+
+You can leave these settings blank and have reasonable defaults be chosen for you, or you can express the models explicitly.
+
+```json
+  ...
+  "modelConfiguration": {
+    "defaultModels": {
+      "chat": "anthropic::2023-06-01::claude-3.5-sonnet",
+      "codeCompletion": "anthropic::2023-06-01::claude-3.5-sonnet",
+      "fastChat": "anthropic::2023-06-01::claude-3-haiku"
+    }
+  }
+  ...
+```
+
+The format of these strings is a "Model Reference", which is a format for uniquely identifying each LLM model exposed from your Sourcegraph instance.
+
+## Advanced Configruation
+
+For most administrators, relying on the LLM models made available by Cody Gateway is sufficient. However, if even more customization is required, you can configure your own LLM providers and models.
+
+<Callout type="warning">Defining your own LLM providers and models is an advanced use-case and requires care to get the correct results. It also may bypass protections to ensure compatibiloity between your Sourcegraph instance and LLMs. If you need help contact your Sourcegraph account executive.</Callout>
+
+### Overview
+
+The `"modelConfiguration"` section exposes two fields `"providerOverrides"` and `"modelOverrides"`. These may override any Sourcegraph-supplied data, or simply introduce new ones entirely.
+
+### Provider Configuration
+
+A "provider" a way to organize LLM models. Typically a provider would be referring to the company that produced the model. Or the specific API/service being used to access the model. But conceptually, it's just a namespace.
+
+By defining a provider override in your Sourcegraph site configuration, you are introducing a new namespace to contain models. Or customize the existing provider namespace supplied by Sourcegraph. (e.g. all `"anthropic"` models.)
+
+The following configuration shippet defines a single provider override with the ID `"anthropic-direct"`.
+
+```json
+"modelConfiguration": {
+  // Do not use any Sourcegraph-supplied models.
+  "sourcegraph": null,
+
+  // Introduce our own, custom provider "anthropic-direct".
+  "providerOverrides": [
+    {
+      "id": "anthropic-direct",
+      "displayName": "Anthropic models, sent directly to anthropic.com",
+
+      // The server-side config section defines how this provider operates.
+      "serverSideConfig": {
+        "type": "anthropic",
+        "accessToken": "sk-ant-api03-xxxxxxxxx",
+        "endpoint": "https://api.anthropic.com/v1/messages"
+      },
+
+      // The default model configuration provides defaults for all LLM
+      // models using this provider.
+      "defaultModelConfig": {
+        "capabilities": [
+          "chat",
+          "autocomplete"
+        ],
+        "contextWindow": {
```

Comment:
@slimsag How does smartContextWindow come into all this configuration btw? Do I get it if I don't change the contextWindow on the models that are supported?

Reply:
Currently, if you are using the new modelconfig version of site config, we set smartContext to "disabled". See https://github.com/sourcegraph/sourcegraph/blob/e5178a6bc0e0f2f5208f9dedb0c226661c881d98/cmd/frontend/internal/modelconfig/resolver.go#L94-L107.

I'm not sure what the ramifications of that is, since we are passing the more specialized context window sizes down to the cody client with the new model metadata. (So rather than _only_ using `[siteconfig]completions.maxChatTokens` we instead have a per-model value, defined in the configuration here.)

Does that all sound right? If using modelconfig should we default smartContext to "enabled"?

Reply:
It's probably best to leave that enabled by default. The Client side logic will take care of whether the models are supported by it or not (can you confirm this @abeatrix ?)

Reply:
@jamesmcnamara is probably the best person to answer this for the client side

Reply:
I am not familiar with the smart context window, but from what I can see I see most of the uses related to smart context window are for old style models and the few references I see to the models specific context window are don't intersect with smart context feature enablement (i.e. they don't seem to care if that flag is enabled or not they just get the models context window).

But I am not familiar with this issue so I could be wrong, there might be some sneaky call at the top of the tree that does actually care about that setting.